src/lib.rs

```diff
@@ -0,0 +1,534 @@
+use {
+    memtest::{MemtestError, MemtestOutcome, MemtestType},
+    prelude::*,
+    rand::{seq::SliceRandom, thread_rng},
+    std::{
+        fmt,
+        time::{Duration, Instant},
+    },
+};
+
+mod memtest;
+mod prelude;
+
+#[derive(Debug)]
+pub struct Memtester {
+    test_types: Vec<MemtestType>,
+    timeout: Duration,
+    require_memlock: bool,
+    allow_working_set_resize: bool,
+    allow_mem_resize: bool,
+    allow_multithread: bool,
+    allow_early_termination: bool,
+}
+
+// TODO: Replace MemtesterArgs with a Builder struct implementing fluent interface
+/// A set of arguments that define the behavior of Memtester
+#[derive(Debug)]
+pub struct MemtesterArgs {
+    /// How long should Memtester run the test suite before timing out
+    pub timeout: Duration,
+    /// Whether memory will be locked before testing
+    /// If memory locking failed but is required, Memtester returns with error
+    pub require_memlock: bool,
+    /// Whether the process working set can be resized to accomodate memory locking
+    /// This argument is only meaninful for Windows
+    pub allow_working_set_resize: bool,
+    /// Whether the requested memory size of testing can be reduced to accomodate memory locking
+    /// This argument is only meaninful is memlock is required
+    pub allow_mem_resize: bool,
```

Comment:
As commented, `allow_mem_resize` is only meaningful (I notice there's a typo as I am typing this😅) when `require_memlock` is true. I am wondering if a 3-way enum (`MemLockFixSize`, `MemLockResizable`, `NoMemLock`) is better for this case? But I'm also unsure if it's troublesome for users to import an additional enum just for something relatively minor

Reply:
I think you're right -- When you have things like this (2 `bool`s but only 3/4 combinations are meaningful), an enum might be a better idea! Don't worry about your user having to import more stuff -- You're helping them make their code more clear as well :)